source/common/common/thread.cc

```diff
@@ -1,36 +1,99 @@
 #include "source/common/common/thread.h"
 
+#include <thread>
+
+#include "source/common/common/assert.h"
+#include "source/common/common/macros.h"
+
 namespace Envoy {
 namespace Thread {
 
-bool MainThread::isMainThread() {
-  // If threading is off, only main thread is running.
-  auto main_thread_singleton = MainThreadSingleton::getExisting();
-  if (main_thread_singleton == nullptr) {
-    return true;
+namespace {
+
+// Singleton structure capturing which thread is the main dispatcher thread, and
+// which is the test thread. This info is used for assertions around catching
+// exceptions and accessing data structures which are not mutex-protected, and
+// are expected only from the main thread.
+//
+// TODO(jmarantz): avoid the singleton and instead have this object owned
+// by the ThreadFactory. That will require plumbing the API::API into all
+// call-sites for isMainThread(), which might be a bit of work, but will make
+// tests more hermetic.
+struct ThreadIds {
+  // Determines whether we are currently running on the main-thread or
+  // test-thread. We need to allow for either one because we don't establish
+  // the full threading model in all unit tests.
+  bool inMainOrTestThread() const {
+    std::thread::id id = std::this_thread::get_id();
+    absl::MutexLock lock(&mutex_);
+    return ((main_thread_use_count_ != 0 && main_thread_id_ == id) ||
+            (test_thread_use_count_ != 0 && test_thread_id_ == id));
+  }
+
+  // Returns a singleton instance of this. The instance is never freed.
+  static ThreadIds& get() { MUTABLE_CONSTRUCT_ON_FIRST_USE(ThreadIds); }
+
+  // Call this when the MainThread exits. Nested semantics are supported, so
+  // that if multiple MainThread instances are declared, we unwind them
```

Comment:
How do we end up with multiple simultaneous main threads?

Reply:
One of the key changes in this PR is that it requires that the thread context be setup before calling inMainOrTestThread(). In the PR, to make that guarantee most easily, this is done redundantly in every main(), and also (for unit tests) in the ThreadLocal constructor. So we can redeclare the same thread in some cases.

Prior to this PR if no one declared main-thread or test-thread, then that test would pass by default.